### ISSUE TITLE
ntpd can't enter SYNCED state

### ISSUE THREAD

**Issue by LynxChaus** [openntpd-portable/openntpd-portable#65]
https://github.com/openntpd-portable/openntpd-portable/blob/b3be9b9269b80d2574dde155316211cb4d12c66e/patches/0014-use-adjtimex-over-adjtime-check-return-value-and-adj.patch#L59
This condition never match - adjtimex() always return STA_UNSYNC with tx.offset = 0 when synced = 0, but STA_UNSYNC reset in update_time_sync_status() which never happen - synced is not set to 1.

**Comment by busterb** [comment]
Confirmed, thanks. I removed the setting of STA_UNSYNC in update_time_sync_status() and it syncs of course. I'm thinking of making ntpd_adjtime more os-specific, since the current logic is hard to follow between ifdefs and compat code.

**Comment by auv-ee** [comment]
I have also encountered this issue.  For reference I'm working on an old, embedded linux-2.6.20gum OS that I can't upgrade.  

do_adjtimex() in linux-2.6.20gum/kernel/time/ntp.c returns TIME_ERROR if  (STA_UNSYNC|STA_CLOCKERR).  The code for openNTPD in openntpd-6.8p1/src/ntpd.c, ntpd_adjtime(), never sets synced=1 with any non-zero TIME_ERROR, but only reports an error if the TIME_ERROR contains a bit other than STA_UNSYNC.  I fixed the problem by changing these two lines:

```
        if (rc == TIME_ERROR) {
                if ((tx.status & ~STA_UNSYNC))
```
into:
```
       if (rc == TIME_ERROR && (tx.status & ~STA_UNSYNC))
```

**Comment by auv-ee** [comment]
I should add that I made the change to the if statements in openntpd-6.8p1/src/ntpd.c, WITHOUT making any change to update_time_sync_status(). The change to ntpd_adjtime() causes the clock to sync only after it is determined that it should (remaining adjustment returned from adjtimex() is zero); while I think never setting STA_UNSYNC in update_time_sync_status() makes the clock always appear synced, whether it is or not (or is this overridden elsewhere?).

**Comment by busterb** [comment]
Thanks for the updates here @auv-ee.

I'd like to propose just removing the entire ADJTIMEX path from this function and leaving it using adjtime(3) from upstream. I'm curious if this works as expected on your system as well. I believe I inherited this original patch from a much older ntpd portable version, but I'm not sure what problem it was actually solving, so if removing it works too, I'd be fine leaving it as-is and being closer to upstream (and one patch less to maintain). This may solve some other issues as well.

**Comment by busterb** [comment]
I just updated master to build with the recently-tagged OPENBSD_6_9 branch, and removed this patch entirely. Please give it a try and let me know what needs to change, if anything.

**Comment by auv-ee** [comment]
Standby.  I'll try to reply to this within a week.  I have to finish another project first.

**Comment by Ypnose** [comment]
Is there a follow-up ? All my systems using `6.8p1` never get synced.

**Comment by auv-ee** [comment]
Just to let busterb know: 
I said I would get back about his proposed resolution within a week.  It is now 6 months out, and I'm sorry to say that I will not be testing this again.  Now *almost* fully retired and no longer have access the equipment.  Sorry.

Ypnose:
It is possible that the simple changes I made to openntpd-6.8p1/src/ntpd.c and reported on 2021/2/22, above, will work for you.

**Comment by Ypnose** [comment]
Thanks. Le'ts hope a new release will be also launched, especially after the recent OpenBSD 7.0.

**Comment by freshprince** [comment]
Will `6.9p1` be released officially on https://www.openntpd.org/portable.html anytime soon? I'm using it on my machines for quite a while without any issues. I had to update because of this problem but would like to move back to the package from the distribution but they won't update as long as there's no official release.

**Comment by dcepelik** [comment]
Hi! Today I noticed that one of my machines running 6.8p1 (portable) initially starts with time delayed by ~20 minutes. It quickly gets synced by OpenNTPd, but the time never gets written back to the RTC. I have eventually debugged this to `STA_UNSYNC` being set in the status field of `struct timex` retrieved by `adjtimex`, and I found this issue, which I think is the culprit. (When I manually clear `STA_UNSYNC` and set `maxerror = 0`, Linux will correctly store the system clock to the RTC every ~11 minutes as expected.)

Is there anything I can do to help with the release?

**Comment by arachsys** [comment]
For what it's worth, I'm running the 7.2 version from the HEAD of the github master branch on a (musl) linux system and it works well. Dropping that complex adjtimex patch did indeed fix the sync problem.

I actually do still patch in adjtimex() support, but in a simpler way than the patch that was dropped: in an adjtime()-like single-shot mode. The point of doing this is that musl adjtime() returns EINVAL on a smaller max offset than the largest offset OpenBSD supports, returning the error instead of saturating at the max value.

My patch looks something like this:
```diff
diff --git a/src/ntpd.c b/src/ntpd.c
index ebd86b6..68c61cc 100644
--- a/src/ntpd.c
+++ b/src/ntpd.c
@@ -17,6 +17,9 @@
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  */
 
+#ifdef HAVE_ADJTIMEX
+#include <sys/timex.h>
+#endif
 #include <sys/types.h>
 #include <sys/resource.h>
 #include <sys/socket.h>
@@ -515,21 +518,34 @@ reset_adjtime(void)
 int
 ntpd_adjtime(double d)
 {
-	struct timeval	tv, olddelta;
 	int		synced = 0;
 	static int	firstadj = 1;
+	double threshold = (double)LOG_NEGLIGIBLE_ADJTIME / 1000;
 
 	d += getoffset();
-	if (d >= (double)LOG_NEGLIGIBLE_ADJTIME / 1000 ||
-	    d <= -1 * (double)LOG_NEGLIGIBLE_ADJTIME / 1000)
+	if (d >= threshold || d <= -1 * threshold)
 		log_info("adjusting local clock by %fs", d);
 	else
 		log_debug("adjusting local clock by %fs", d);
+
+#ifdef HAVE_ADJTIMEX
+	struct timex tx = { 0 };
+	tx.offset = d * 1000000;
+	tx.modes = ADJ_OFFSET_SINGLESHOT;
+
+	if (adjtimex(&tx) < 0) {
+		log_warn("adjtimex failed");
+	} else if (!firstadj && tx.offset == 0) {
+		synced = 1;
+	}
+#else
+	struct timeval	tv, olddelta;
 	d_to_tv(d, &tv);
 	if (adjtime(&tv, &olddelta) == -1)
 		log_warn("adjtime failed");
 	else if (!firstadj && olddelta.tv_sec == 0 && olddelta.tv_usec == 0)
 		synced = 1;
+#endif
 	firstadj = 0;
 	update_time_sync_status(synced);
 	return (synced);
```
I don't know whether it would be needed on glibc systems - they may not impose the same hard-limit with an error.

**Comment by perkelix** [comment]
@busterb in which release was this fixed?

**Comment by busterb** [comment]
That was a while back, but it the changes would have shipped with 6.9, but it was unreleased. I like @arachsys's straightforward approach though.

**Comment by busterb** [comment]
This will be fixed in the next release. I've been running a later version of the above code for a couple of weeks without issue.
